contrib/endpoints/src/api_manager/config_manager.cc

```diff
@@ -0,0 +1,163 @@
+  }
+
+  service_management_fetch_.reset(new ServiceManagementFetch(global_context));
+}
+
+void ConfigManager::Init() {
+  if (global_context_->service_name().empty() ||
+      global_context_->config_id().empty()) {
+    GlobalFetchGceMetadata(global_context_, [this](utils::Status status) {
+      OnFetchMetadataDone(status);
+    });
+  } else {
+    OnFetchMetadataDone(utils::Status::OK);
+  }
+}
+
+void ConfigManager::OnFetchMetadataDone(utils::Status status) {
+  if (!status.ok()) {
+    // We should not get here
+    global_context_->env()->LogError("Unexpected status: " + status.ToString());
+    rollout_apply_function_(utils::Status(Code::ABORTED, status.ToString()),
+                            kEmptyConfigs);
+    return;
+  }
+
+  // Update service_name
+  if (global_context_->service_name().empty()) {
+    global_context_->set_service_name(
+        global_context_->gce_metadata()->endpoints_service_name());
+  }
+
+  // Update config_id
+  if (global_context_->config_id().empty()) {
+    global_context_->config_id(
+        global_context_->gce_metadata()->endpoints_service_config_id());
+  }
+
+  // Call ApiManager with status Code::ABORTED, ESP will stop moving forward
+  if (global_context_->service_name().empty()) {
+    std::string msg = "API service name is not specified";
+    global_context_->env()->LogError(msg);
+    rollout_apply_function_(utils::Status(Code::ABORTED, msg), kEmptyConfigs);
+    return;
+  }
+
+  // TODO(jaebong) config_id should not be empty for the first version
+  // This part will be removed after the rollouts feature added
+  if (global_context_->config_id().empty()) {
+    std::string msg = "API config_id is not specified";
+    global_context_->env()->LogError(msg);
+    rollout_apply_function_(utils::Status(Code::ABORTED, msg), kEmptyConfigs);
+    return;
+  }
+
+  // Fetch service account token
+  GlobalFetchServiceAccountToken(global_context_, [this](utils::Status status) {
+    OnFetchAuthTokenDone(status);
+  });
+}
+
+void ConfigManager::OnFetchAuthTokenDone(utils::Status status) {
+  if (!status.ok()) {
+    // We should not get here
+    global_context_->env()->LogError("Unexpected status: " + status.ToString());
+    rollout_apply_function_(utils::Status(Code::ABORTED, status.ToString()),
+                            kEmptyConfigs);
+    return;
+  }
+
+  // Fetch configs from the Inception
+  // For now, config manager has only one config_id (100% rollout)
+  std::shared_ptr<ConfigsFetchInfo> config_fetch_info(
+      new ConfigsFetchInfo({{global_context_->config_id(), 100}}));
+
+  FetchConfigs(config_fetch_info);
+}
+
+// Fetch configs from rollouts. fetch_info has rollouts and fetched configs
+void ConfigManager::FetchConfigs(
+    std::shared_ptr<ConfigsFetchInfo> config_fetch_info) {
+  for (auto rollout : config_fetch_info->rollouts) {
+    service_management_fetch_->GetConfig(
+        rollout.first, rollout.second,
+        [this, config_fetch_info](const utils::Status& status,
```

Comment:
config_id and percentage can be pass to lambda capture,  not to the GetConfig() function,  that function doesn't need them.